pkg/upgrade/upgrader/jiva_upgrade.go

```diff
@@ -633,3 +683,225 @@ func jivaUpgrade(pvName, openebsNamespace string, utaskObj *utask.UpgradeTask) (
 	klog.Info("Upgrade Successful for", pvName)
 	return options.utaskObj, nil
 }
+
+func (j *jivaVolumeOptions) cleanup(openebsNamespace string) error {
+	var err error
+	if j.replicaObj.version == currentVersion && j.replicaObj.name != "" {
+		klog.Info("cleaning old replica deployment")
+		err = deployClient.WithNamespace(j.ns).Delete(j.replicaObj.name, &metav1.DeleteOptions{})
+		if err != nil {
+			return err
+		}
+	}
+	if j.controllerObj.version == currentVersion && j.ns != openebsNamespace {
+		klog.Info("cleaning old replica deployment")
+		err = deployClient.WithNamespace(j.ns).Delete(j.controllerObj.name, &metav1.DeleteOptions{})
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (j *jivaVolumeOptions) migrate(pvName, openebsNamespace string) error {
+	var err error
+	// if old replica is missing then migration was
+	// successful till replica cleanup in previous iteration
+	if j.replicaObj.name != "" {
+		err = j.migrateReplica(openebsNamespace)
+		if err != nil {
+			return err
+		}
+	}
+	// if pvc deployed in openebs namespace no need
+	// to migrate the controller deployment
+	if j.ns != openebsNamespace {
+		err = j.migrateTarget(pvName, openebsNamespace)
+	}
+	return err
+}
+
+func getNodeNames(deployObj *appsv1.Deployment) (int, []string) {
+	matchExp := deployObj.Spec.Template.Spec.Affinity.NodeAffinity.
+		RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms[0].MatchExpressions
+	for i, exp := range matchExp {
+		if exp.Key == "kubernetes.io/hostname" {
+			return i, exp.Values
+		}
+	}
+	return -1, nil
+}
+
+func (j *jivaVolumeOptions) migrateReplica(openebsNamespace string) error {
+	// get the old replica deployment by name
+	oldDeployObj, err := deployClient.WithNamespace(j.ns).Get(j.replicaObj.name)
+	if err != nil {
+		return err
+	}
+	index, nodeNames := getNodeNames(oldDeployObj)
+	if index == -1 {
+		return errors.New("unable to find kubernetes.io/hostname key in nodeAffinity")
+	}
+	replicaCount := len(nodeNames)
+	// get the separate replica deployments in openebs namespace
+	deployList, err := deployClient.WithNamespace(openebsNamespace).List(&metav1.ListOptions{
+		LabelSelector: replicaDeployLabel,
+	})
+	if err != nil {
+		return err
+	}
+	replicasCreated := len(deployList.Items)
+	// if the volume was deployed in openebs namespace while provisioning
+	// as the old deployment also has the same label
+	if j.ns == openebsNamespace {
+		replicasCreated = replicasCreated - 1
+	}
+
+	klog.Infof("splitting replica deployment")
+	var zero int32
+	for i := replicasCreated; i < replicaCount; i++ {
```

Comment:
One comment above this for loop to say that:
replica deployment `pv-name-rep` will be split into multiple replicas like `pv-name-rep-1`, `pv-name-rep-2`,... `pv-name-rep-n`, where n is the replica count for this volume.